addons/point_of_sale/models/pos_session.py

```diff
@@ -1842,314 +1950,59 @@ def _ensure_access_token(self):
-            },
-        }
-
-    def _get_pos_ui_pos_payment_method(self, params):
-        return self.env['pos.payment.method'].search_read(**params['search_params'])
-
-    def _loader_params_account_fiscal_position(self):
-        return {'search_params': {'domain': [('id', 'in', self.config_id.fiscal_position_ids.ids)], 'fields': []}}
-
-    def _get_pos_ui_account_fiscal_position(self, params):
-        return self.env['account.fiscal.position'].search_read(**params['search_params'])
-
-    def _loader_params_account_fiscal_position_tax(self):
-        loaded_data = self._context.get('loaded_data')
-        fps = loaded_data['account.fiscal.position']
-        fiscal_position_tax_ids = sum([fpos['tax_ids'] for fpos in fps], [])
-        return {'search_params': {'domain': [('id', 'in', fiscal_position_tax_ids)], 'fields': []}}
-
-    def _get_pos_ui_account_fiscal_position_tax(self, params):
-        return self.env['account.fiscal.position.tax'].search_read(**params['search_params'])
-
-    def get_pos_ui_product_product_by_params(self, custom_search_params):
-        """
-        :param custom_search_params: a dictionary containing params of a search_read()
-        """
-        params = self._loader_params_product_product()
-        # custom_search_params will take priority
-        params['search_params'] = {**params['search_params'], **custom_search_params}
-        products = self.env['product.product'].with_context(active_test=False).search_read(**params['search_params'])
-        if len(products) > 0:
-            self._process_pos_ui_product_product(products)
-        return products
+                product['lst_price'] = self.company_id.currency_id._convert(product['lst_price'], self.config_id.currency_id, self.company_id, fields.Date.today())
 
     def get_pos_ui_res_partner_by_params(self, custom_search_params):
         """
         :param custom_search_params: a dictionary containing params of a search_read()
         """
-        params = self._loader_params_res_partner()
+        params = self._load_data_params(self.config_id)['search_read']['res.partner']
         # custom_search_params will take priority
-        params['search_params'] = {**params['search_params'], **custom_search_params}
-        partners = self.env['res.partner'].search_read(**params['search_params'])
+        params = {**params, **custom_search_params}
+        partners = self.env['res.partner'].search_read(**params)
         return partners
 
     def find_product_by_barcode(self, barcode):
+        load_data_params = self._load_data_params(self.config_id)
         product = self.env['product.product'].search([
             ('barcode', '=', barcode),
             ('sale_ok', '=', True),
             ('available_in_pos', '=', True),
         ])
         if product:
-            return {'product_id': [product.id]}
+            return {'product.product': product.read(load_data_params['search_read']['product.product']['fields'], load=False)}
```

Comment:
We don't need this kind of strict data structure for rpc methods, it's very restricting. The developer should be allowed to just return any value.

Reply:
Depending on which data_service function you will use, these is one who's allow you to return everything u want